src/cdk/tree/tree.ts

```diff
@@ -100,19 +100,25 @@ export class CdkTreeNode<T>  implements FocusableOption, OnDestroy {
     this._elementRef.nativeElement.focus();
   }
 
-  private _setRoleFromData(): void {
+  protected _setRoleFromData(): void {
     if (this._tree.treeControl.isExpandable) {
       this.role = this._tree.treeControl.isExpandable(this._data) ? 'group' : 'treeitem';
     } else {
       if (!this._tree.treeControl.getChildren) {
         throw getTreeControlFunctionsMissingError();
       }
-      this._tree.treeControl.getChildren(this._data).pipe(takeUntil(this._destroyed))
-        .subscribe(children => {
-          this.role = children && children.length ? 'group' : 'treeitem';
-        });
+      const childrenNodes = this._tree.treeControl.getChildren(this._data);
+      if (Array.isArray(childrenNodes)) {
+        this._setRoleFromChildren(childrenNodes as T[]);
+      } else if (childrenNodes instanceof Observable) {
+        childrenNodes.pipe(takeUntil(this._destroyed)).subscribe(this._setRoleFromChildren);
```

Comment:
You might be getting an incorrect `this` inside `_setRoleFromChildren` by passing in the function like this.